main.cpp

```diff
@@ -24,31 +45,101 @@ int main()
     }
 
     glfwMakeContextCurrent(window);
+    glfwSetKeyCallback(window, KeyCallback);
+
+    // Uncomment this if you would like to blow up the GPU :D
+    // glfwSwapInterval(0);
 
     if (glewInit() != GLEW_OK)
     {
         std::cout << "glewInit() failure" << std::endl;
         return -1;
     }
 
+    GLCall(glEnable(GL_BLEND));
+    GLCall(glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA));
+
     std::cout << "Using OpenGL version: " << glGetString(GL_VERSION) << std::endl;
 
+    int numTextureSlots;
+    glGetIntegerv(GL_MAX_TEXTURE_IMAGE_UNITS, &numTextureSlots);
+    std::cout << "Texture Slots Available: " << numTextureSlots << std::endl;
+
+    Shader shader = Shader("D:\\code\\pong\\basic.shader");
+    Renderer renderer;
+
+    player = new Rectangle(25.0f, 225.0f);
```

Comment:
Prefer unique or shared pointer "no raw new"

```
player = make_unique<Rectangle>(25.0f, 225.0f);
```

Choose shared or unique based on how many owners it will have - 1 or many.

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#rsmart-smart-pointers

Reply:
No more manual refcounting for me!!!
Interestingly, when I was trying to convert the mTexture variable on Rectangle to a smart pointer, I was getting this cryptic conversion warning:
```
warning C4244: 'argument': conversion from '_Ty' to  
'unsigned char', possible loss of data
          with
          [
              _Ty=int
          ]
```
 The constructor for the SolidColorTexture I was using takes 4 unsigned chars. I was putting 255 for all, and this was causing them to be interepreted as int. Very interestingly though, depending on how I created the smart pointer was what triggered the warning
`std::unique_ptr<SolicColorTexture> temp = std::make_unique<SolidColorTexture>(255, 255, 255, 255);`
did trigger the warning
`std::unique_ptr<SolicColorTexture> temp(new SolidColorTexture(255, 255, 255, 255)`
did not trigger the warning

I did manage to fix it by just using a specified type.
```
const unsigned char kColorValue = 255;
mTexture = std::make_unique<SolidColorTexture>(kColorValue, kColorValue, kColorValue, kColorValue);
```
Seems strange

Reply:
Very interesting... 

I think the way you did it is OK. They basically want to make sure that your number can fit into a char. Funny because you would hope the warnings would be smarter and determine the literal was <=255.

I think the way you did it is perfectly fine :) 

as for your examples, another good use of auto to save you some clutter:

```
auto temp = std::make_unique<SolidColorTexture>(255, 255, 255, 255);
```